sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotHardcodeCredentialsBase.cs

```diff
@@ -0,0 +1,127 @@
+
+        private string credentialWords;
+        private IEnumerable<string> splitCredentialWords;
+        private Regex passwordValuePattern;
+        
+        [RuleParameter("credentialWords", PropertyType.String,
+            "Comma separated list of words identifying potential credentials", DefaultCredentialWords)]
+        public string CredentialWords
+        {
+            get => this.credentialWords;
+            set
+            {
+                this.credentialWords = value;
+                this.splitCredentialWords = value.ToUpperInvariant()
+                    .Split(new[] { ',' }, StringSplitOptions.RemoveEmptyEntries)
+                    .Select(x => x.Trim())
+                    .ToList();
+                this.passwordValuePattern = new Regex(string.Format(@"\b(?<password>{0})\b[:=]\S",
+                    string.Join("|", this.splitCredentialWords)), RegexOptions.Compiled | RegexOptions.IgnoreCase);
+            }
+        }
+
+        public DoNotHardcodeCredentialsBase()
+        {
+            CredentialWords = DefaultCredentialWords;
+        }
+
+        protected abstract class BannedWordsFinderBase<TSyntaxNode>
+             where TSyntaxNode : SyntaxNode
+        {
+            private readonly DoNotHardcodeCredentialsBase analyzer;
+
+            protected BannedWordsFinderBase(DoNotHardcodeCredentialsBase analyzer)
+            {
+                this.analyzer = analyzer;
+            }
+
+            protected abstract bool IsAssignedWithStringLiteral(TSyntaxNode syntaxNode, SemanticModel semanticModel);
+
+            protected abstract string GetVariableName(TSyntaxNode syntaxNode);
+
+            protected abstract string GetAssignedValue(TSyntaxNode syntaxNode);
+
+            public Action<SyntaxNodeAnalysisContext> GetAnalysisAction(DiagnosticDescriptor rule) =>
+                context =>
+                {
+                    var declarator = (TSyntaxNode)context.Node;
+
+                    if (!IsAssignedWithStringLiteral(declarator, context.SemanticModel))
+                    {
+                        return;
+                    }
+
+                    var variableName = GetVariableName(declarator);
+                    var variableValue = GetAssignedValue(declarator);
+
+                    var bannedWords = FindBannedWords(variableName, variableValue);
+                    if (bannedWords.Any())
+                    {
+                        context.ReportDiagnosticWhenActive(
+                            Diagnostic.Create(rule, declarator.GetLocation(), string.Join(", ", bannedWords)));
+                    }
+                };
+
+            protected IEnumerable<string> FindBannedWords(string variableName, string variableValue)
+            {
+                if (string.IsNullOrWhiteSpace(variableName) ||
+                    string.IsNullOrWhiteSpace(variableValue))
+                {
+                    return Enumerable.Empty<string>();
```

Comment:
could you add a test to cover this line? probably an empty string value?

Reply:
Apparently the check for `variableName` is not needed... I put it because we don't support assignments like `this.passwords = "foo"` and I was getting NRE in a previous step in the execution, but apparently `variableName` is ok to be null, because `SplitCamelCaseToWords` will return empty enumerable...